File: Data Structures & Algorithms/permutations/submission-4.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        '''
        used = [False] * len(nums)
        res, path = [], []

        def backtrack():
            if len(path) == len(nums):
                res.append(path.copy())
                return

            for i in range(len(nums)):
                if not used[i]:
                #Allow append into path
                    path.append(nums[i])
                    used[i] = True
                    backtrack()
                    path.pop()
                    used[i] = False
        backtrack()
        return res

#TimeCP: O(n*n!)
#SpaceCP: O(n*n!) include list ouput.
'''
        if len(nums) == 0:
            return [[]]
        
        perms = self.permute(nums[1:])
        res = []
        for p in perms:
            for i in range(len(p) + 1): #vị trí cần chèn
                tmp = p.copy()
                tmp.insert(i, nums[0])
                res.append(tmp)
        return res 

File: Data Structures & Algorithms/subsets/submission-4.cpp
class Solution {
public:

    vector<vector<int>> res;
    vector<int> path;

    void backtrack(int start, vector<int>& nums) {
        //save path current into res
        res.push_back(path);

        for (int i = start; i < nums.size(); ++i) {
            path.push_back(nums[i]);
            backtrack(i + 1, nums);
            path.pop_back();
        }
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        backtrack(0, nums);
        return res;

    }
};
